### PR TITLE
refactor: Decouple Web Server from Lightning Connection

### PR DESCRIPTION
**Pull request by m1sterc001guy**
Working towards: https://github.com/fedimint/fedimint/issues/2786

Currently, the gateway's webserver is tied to the connection status of the lightning node. This is because there are some operations that the webserver needs a lightning connection to do (like fetching route hints). However, this makes it difficult to allow for "configuration" APIs that might be needed before the lightning connection is even established.

This PR decouples the webserver from the lightning connection. The webserver is now long-lived, but has a `GatewayState` struct that allows it to be in different states. Some APIs will not be accessible when the Gateway is in certain states (i.e `GatewayState::Disconnected`). In the future, we can add states like `Configuring` to allow for a configuration phase before connecting to the lightning node. This also allows us to cache Lightning connection data structures in the `Running` phase.

### CODE REVIEW

File: fedimint-testing/src/gateway.rs [hunk 1]
```diff
@@ -86,46 +87,59 @@ impl GatewayTest {
...
 
-        let listening_addr = lightning.listening_address();
-        let info = lightning.info().await.unwrap();
-
-        let mut tg = TaskGroup::new();
-        // Create the stream to route HTLCs. We cannot create the Gateway until the
-        // stream to the lightning node has been setup.
-        let (stream, ln_client) = lightning.route_htlcs(&mut tg).await.unwrap();
+        let lightning_builder: Arc<dyn LightningBuilder + Send + Sync> =
+            match Fixtures::is_real_test() {
```
Comment (douglaz) [inline fedimint-testing/src/gateway.rs:91]: if/else is much better than match true/false

File: fedimint-testing/src/gateway.rs [hunk 2]
```diff
@@ -86,46 +87,59 @@ impl GatewayTest {
...
+            {
+                break;
+            }
+
+            if gateway_state_iterations > 9 {
+                panic!("Gateway did not start running after 10 attempts");
+            }
+
+            gateway_state_iterations += 1;
+            sleep(Duration::from_millis(100)).await;
```
Comment (douglaz) [inline fedimint-testing/src/gateway.rs:138]: This is a very low waiting period, it will fail on CI

File: gateway/ln-gateway/src/lib.rs [hunk 3]
```diff
@@ -251,30 +310,22 @@ impl Gatewayd {
...
-                                let task_group = run_webserver(self.password.clone(), self.listen, gateway)
-                                    .await
-                                    .expect("Failed to start webserver");
-                                info!("Successfully started webserver");
-
-                                Gateway::handle_htlc_stream(stream, ln_client, handle.clone(), scid_to_federation.clone(), clients.clone()).await;
-                                warn!("HTLC Stream Lightning connection broken. Stopping webserver...");
-                                task_group.shutdown();
+                                let (lightning_public_key, lightning_alias) =
+                                Self::fetch_lightning_node_info(ln_client.clone()).await.expect("Failed to retrieve Lightning info after establishing a connection");
```
Comment (douglaz) [inline gateway/ln-gateway/src/lib.rs:314]: Perhaps retry this and other operations below? But this could be another PR

File: gateway/ln-gateway/src/lib.rs [hunk 4]
```diff
@@ -860,6 +492,140 @@ impl Gateway {
...
+        payload: ConnectFedPayload,
+    ) -> Result<FederationInfo> {
+        if let GatewayState::Running {
+            lnrpc,
+            lightning_public_key,
+            lightning_alias: _,
+        } = self.state.read().await.clone()
+        {
+            let invite_code = InviteCode::from_str(&payload.invite_code).map_err(|e| {
+                GatewayError::InvalidMetadata(format!("Invalid federation member string {e}"))
```
Comment (douglaz) [inline gateway/ln-gateway/src/lib.rs:553]: Use the debug format so the stack trace will be included
```suggestion
                GatewayError::InvalidMetadata(format!("Invalid federation member string {e:?}"))
```

File: gateway/ln-gateway/src/lib.rs [hunk 5]
```diff
@@ -873,4 +639,250 @@ impl Gateway {
...
+                                        .await
+                                        .is_err()
+                                    {
+                                        error!("Error registering federation {federation_id}");
+                                    }
+                                }
+                            }
+                            Err(_) => {
+                                error!(
+                                    "Could not retrieve route hints, gateway will not be registered."
```
Comment (douglaz) [inline gateway/ln-gateway/src/lib.rs:756]: include the error?

File: gateway/ln-gateway/src/lib.rs [hunk 6]
```diff
@@ -873,4 +639,250 @@ impl Gateway {
...
+                                        .register_with_federation(
+                                            api.clone(),
+                                            route_hints.clone(),
+                                            GW_ANNOUNCEMENT_TTL,
+                                            gateway_id,
+                                        )
+                                        .await
+                                        .is_err()
+                                    {
+                                        error!("Error registering federation {federation_id}");
```
Comment (douglaz) [inline gateway/ln-gateway/src/lib.rs:750]: Include the error?

File: fedimint-testing/src/gateway.rs (hunk 2 above)
Comment (m1sterc001guy) [reply fedimint-testing/src/gateway.rs]: I increased it to 10 seconds, waiting a second each time.

File: gateway/ln-gateway/src/lib.rs (hunk 3 above)
Comment (m1sterc001guy) [reply gateway/ln-gateway/src/lib.rs]: The likelihood of this failing is hopefully small since just above it established the intercepted HTLC stream, but I added the retry anyway.

File: gateway/ln-gateway/src/lib.rs [hunk 7]
```diff
@@ -251,249 +310,82 @@ impl Gatewayd {
...
-
-                                let task_group = run_webserver(self.password.clone(), self.listen, gateway)
-                                    .await
-                                    .expect("Failed to start webserver");
-                                info!("Successfully started webserver");
-
-                                Gateway::handle_htlc_stream(stream, ln_client, handle.clone(), scid_to_federation.clone(), clients.clone()).await;
-                                warn!("HTLC Stream Lightning connection broken. Stopping webserver...");
-                                task_group.shutdown();
+                                if let Ok((lightning_public_key, lightning_alias)) =
```
Comment (douglaz) [inline gateway/ln-gateway/src/lib.rs:313]: `match` here and log the `Err(e)`. Avoid swallowing errors

File: gateway/ln-gateway/src/lib.rs [hunk 8]
```diff
@@ -860,6 +497,140 @@ impl Gateway {
         ))
     }
 
+    async fn handle_pay_invoice_msg(&self, payload: PayInvoicePayload) -> Result<Preimage> {
+        if let GatewayState::Running {
+            lnrpc: _,
+            lightning_public_key: _,
+            lightning_alias: _,
+        } = self.state.read().await.clone()
```
Comment (elsirion) [inline gateway/ln-gateway/src/lib.rs:505]: Could be a let-else statement to avoid nesting.

**Review by elsirion — APPROVED** [review]

File: devimint/src/external.rs [hunk 9]
```diff
@@ -444,10 +444,10 @@ pub enum LightningNode {
 }
 
 impl LightningNode {
-    pub fn name(&self) -> LightningNodeName {
+    pub fn name(&self) -> LightningNodeType {
```
[the remark this answers is not in the document]
Comment (Jjajsja) [reply devimint/src/external.rs:447]: https://github.com/fedimint/fedimint/issues/2786#issue-1812243842
